plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerUpdatesTest.java

```diff
@@ -0,0 +1,133 @@
+package io.jenkins.tools.pluginmanager.impl;
+
+import io.jenkins.tools.pluginmanager.config.Config;
+import io.jenkins.tools.pluginmanager.config.Settings;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import org.apache.commons.io.IOUtils;
+import org.json.JSONObject;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class PluginManagerUpdatesTest {
+
+    private PluginManager pm;
+
+    @Before
+    public void before() throws IOException {
+        Config cfg = Config.builder()
+                .withJenkinsWar(Settings.DEFAULT_WAR)
+                .build();
+
+        pm = new PluginManager(cfg);
+
+        JSONObject pluginInfoJson = loadPluginVersionsFromClassPath();
+        pm.setPluginInfoJson(pluginInfoJson);
+    }
+
+    @Test
+    public void simpleUpdate() {
+        List<Plugin> latestVersionsOfPlugins = pm.getLatestVersionsOfPlugins(singletonList(plugin("mailer", "1.31")));
+
+        assertThat(latestVersionsOfPlugins)
+                .containsExactly(plugin("mailer", "1.32"));
+    }
+
+    private JSONObject loadPluginVersionsFromClassPath() throws IOException {
+        try (InputStream stream = getClass().getResourceAsStream("available-updates/simple-plugin-versions.json")) {
+            return new JSONObject(IOUtils.toString(stream, StandardCharsets.UTF_8));
+        }
+    }
+
+    @Test
+    @Ignore("Need another source of data for incrementals, add later")
+    public void updateIncremental() {
+        List<Plugin> latestVersionsOfPlugins = pm.getLatestVersionsOfPlugins(singletonList(
+                pluginIncremental("mailer", "org.jenkins-ci.plugins", "1.31-rc315.eb08e134da74")
+        ));
+
+        assertThat(latestVersionsOfPlugins)
+                .containsExactly(pluginIncremental("mailer", "org.jenkins-ci.plugins", "1.31-rc316.fb08e134da74"));
+    }
+
+    @Test
+    public void incrementalIsnotUpgradedToGA() {
+        List<Plugin> latestVersionsOfPlugins = pm.getLatestVersionsOfPlugins(singletonList(
+                pluginIncremental("mailer", "org.jenkins-ci.plugins", "1.31-rc316.fb08e134da74")
+        ));
+
+        assertThat(latestVersionsOfPlugins)
+                .containsExactly(pluginIncremental("mailer", "org.jenkins-ci.plugins", "1.31-rc316.fb08e134da74"));
+    }
+
+    @Test
+    public void nonExistentPlugin() {
+        List<Plugin> latestVersionsOfPlugins = pm.getLatestVersionsOfPlugins(singletonList(plugin("non-existing-plugin", "1.31")));
```

Comment:
It seems weird to say that the latest version is valid if the plugin doesn't exist, but I am not sure what else to do.

Reply:
I don't think an update check should fail if it's not there,
not 100% sure how blacklisting behaves but I think that it disappears from the update center but you can still download it.